src/background_music/FrameBasedEffectPlugin.h

```diff
@@ -0,0 +1,190 @@
+ * limitations under the License.
+ */
+#include <libaudcore/plugin.h>
+#include <libaudcore/ringbuf.h>
+
+#include "Detection.h"
+
+template<class D>
+class FrameBasedEffectPlugin : public EffectPlugin
+{
+    static_assert(std::is_base_of_v<Detection, D>);
+
+public:
+    constexpr FrameBasedEffectPlugin(const PluginInfo info, int order)
+        : EffectPlugin(info, order, true)
+    {
+    }
+    virtual ~FrameBasedEffectPlugin() = default;
+
+    bool init() final;
+    void cleanup() final;
+
+    void start(int & channels, int & rate) final;
+    Index<float> & process(Index<float> & data) final;
+    bool flush(bool force) final;
+    Index<float> & finish(Index<float> & data, bool end_of_playlist) final;
+    int adjust_delay(int delay) final;
+    bool offer_frame_return_if_output();
+
+    virtual bool after_init() { return true; }
+    virtual void before_cleanup() {}
+
+    Index<float> frame_in;
+    RingBuf<float> read_ahead_buffer;
+    Index<float> frame_out;
+    Index<float> output;
+    int processed_frames = 0;
+    int current_channels = 0, current_rate = 0, channel_last_read = 0;
+    D * detection = nullptr;
+};
+
+template<class D>
+bool FrameBasedEffectPlugin<D>::init()
+{
+    if (!detection)
+    {
+        detection = new D;
+    }
+    return after_init();
+}
+
+template<class D>
+void FrameBasedEffectPlugin<D>::cleanup()
+{
+    before_cleanup();
+    read_ahead_buffer.destroy();
+    output.clear();
+    frame_in.clear();
+    frame_out.clear();
+    if (detection)
+    {
+        delete detection;
+        detection = nullptr;
+    }
+}
+
+template<class D>
+void FrameBasedEffectPlugin<D>::start(int & channels, int & rate)
+{
+    current_channels = channels;
+    current_rate = rate;
+    channel_last_read = 0;
+    processed_frames = 0;
+
+    detection->start(channels, rate);
+    // Set the initial values for the integrators to the center value
+    // (naturally squared as we integrate squares)
+
+    // As data is added, then fetched, we need 1 extra frame in the buffer.
+    int alloc_size = current_channels * (detection->read_ahead() + 1);
+    if (read_ahead_buffer.size() < alloc_size)
+    {
+        read_ahead_buffer.alloc(alloc_size);
+    }
+
+    frame_in.resize(current_channels);
+    frame_out.resize(current_channels);
+
+    flush(false);
+}
+
+template<class D>
+Index<float> & FrameBasedEffectPlugin<D>::process(Index<float> & data)
+{
+    detection->update_config();
+
+    int output_samples = 0;
+    output.resize(0);
+
+    // It is assumed data always contains a multiple of channels, but we don't
+    // care.
+    for (float sample : data)
+    {
+        frame_in[channel_last_read++] = sample;
+        if (channel_last_read == current_channels)
+        {
+            // Processing happens per frame. Because of read-ahead there is not
+            // always output available yet.
+            if (offer_frame_return_if_output())
```

Comment:
So we're performing a multi-level function call (`offer_frame_return_if_output` calls several more functions) on every single frame of audio? As in, 48,000 (or whatever) times per second? This seems quite sub-optimal. Is it really necessary? Is there no way to process a bit more data per function call?

Reply:
To be honest, I think a lot will be inlined, as the compiler triggers warnings when adding inline keywords. The intention is to move a bit more responsibilities, like the read ahead buffer to the detection class and that might offer some opportunities for optimization.